src/vm/jitinterface.cpp

```diff
@@ -7993,57 +7994,91 @@ CorInfoInline CEEInfo::canInline (CORINFO_METHOD_HANDLE hCaller,
+            {
+                result = INLINE_FAIL;
+                szFailReason = "Profiler disabled inlining globally";
+                goto exit;
+            }
 
-                if (SUCCEEDED(hr) && !fShouldInline)
+#if defined(PROFILING_SUPPORTED) && !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
+            if (CORProfilerEnableRejit())
+            {
+                CodeVersionManager* pCodeVersionManager = pCallee->GetCodeVersionManager();
+                CodeVersionManager::TableLockHolder lock(pCodeVersionManager);
+                ILCodeVersion ilVersion = pCodeVersionManager->GetActiveILCodeVersion(pCallee);
+                if (ilVersion.GetRejitState() != ILCodeVersion::kStateActive || !ilVersion.HasDefaultIL())
                 {
                     result = INLINE_FAIL;
-                    szFailReason = "Profiler disabled inlining locally";
+                    szFailReason = "ReJIT methods cannot be inlined.";
                     goto exit;
                 }
             }
-            END_PIN_PROFILER();
+#endif // defined(PROFILING_SUPPORTED) && !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
+
+            // If the profiler wishes to be notified of JIT events and the result from
+            // the above tests will cause a function to be inlined, we need to tell the
+            // profiler that this inlining is going to take place, and give them a
+            // chance to prevent it.
+            {
+                BEGIN_PIN_PROFILER(CORProfilerTrackJITInfo());
+                if (pCaller->IsILStub() || pCallee->IsILStub())
+                {
+                    // do nothing
+                }
+                else
+                {
+                    BOOL fShouldInline;
+                    HRESULT hr = g_profControlBlock.pProfInterface->JITInlining(
+                        (FunctionID)pCaller,
+                        (FunctionID)pCallee,
+                        &fShouldInline);
+
+                    if (SUCCEEDED(hr) && !fShouldInline)
+                    {
+                        result = INLINE_FAIL;
+                        szFailReason = "Profiler disabled inlining locally";
+                        goto exit;
+                    }
+                }
+                END_PIN_PROFILER();
+            }
         }
+
+#if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
+        // NOTE: this section needs to remain as the last bit of code before the exit block below.
+        // If more checks are added after this we will potentially over report inlinings.
+
+        // Since we exit early for INLINE_FAIL, it always should be INLINE_PASS at this point.
+        _ASSERTE(result == INLINE_PASS);
+        // We don't want to track the chain of methods, so intentionally use m_pMethodBeingCompiled
+        // to just track the methods that pCallee is eventually inlined in
+        pCallee->GetModule()->AddInlining(pOrigCaller, pCallee);
```

Comment:
You should do this in `reportInliningDecision` instead. There is a bunch of checks that the JIT does after this, and so you will be getting a bunch of false positives by recording it here. `reportInliningDecision` is when the inlining decision is done and confirmed.